site/content/entry/yuzu-progress-report-apr-2022/index.md

```diff
@@ -0,0 +1,332 @@
+
+{{< imgs
+	"./coords.png| byte[] provides these great examples"
+  >}}
+
+This isn’t an issue if you want to play Super Mario Galaxy in yuzu with Vulkan, behaviour is 1:1 with what the game expects. 
+But if you want to play using OpenGL instead, the game will give coordinates that are flipped compared to what your GPU will process, being limited by OpenGL’s coordinate system.
+
+The solution is very simple, {{< gh-hovercard "8149" "flip the front faces" >}} when the Z-axis is inverted.
+
+{{< single-title-imgs-compare
+	"Welcome to the Shadow Realm Resort (Super Mario Sunshine)"
+	"./zbug.png"
+	"./zfix.png"
+>}}
+
+Next in line, you may have noticed that `Super Mario Sunshine` rendered with a black bar at the bottom.
+This is because the Wii and GC games natively use an aspect ratio different to the usual 16:9 we’re used to.
+Instead, the games render at a 5:3 aspect ratio. 
+`Super Mario Galaxy` informs the system to crop the screen properly, but `Super Mario Sunshine` doesn't, so yuzu previously had to fill the remaining space with black pixels.
+
+{{< imgs
+	"./crop.png| "
+  >}}
+
+While proportions would be correct with the black bars, that’s not how Nintendo intended the games to be played, so for accuracy’s sake, byte[] stretches the image to match the native 1920x1080 resolution of the Switch, both {{< gh-hovercard "8150" "for Vulkan" >}} and {{< gh-hovercard "8152" "for OpenGL" >}}.
+
+{{< single-title-imgs-compare
+	"Don not adjust your set (Super Mario Sunshine)"
+	"./cropbug.png"
+	"./cropfix.png"
+>}}
+
+In the previous report we mentioned how S8D24 < > ABGR8 texture conversions allow `Super Mario Galaxy` star bits to behave correctly. 
+Well, it’s {{< gh-hovercard "8161" "OpenGL’s turn" >}} to join the fun.
+
+{{< imgs
+	"./s8d24.png| "
+  >}}
+
+We mentioned last month how `Super Mario 64` had special requirements to start running on yuzu. 
+Most games build their code `ahead-of-time` ([AOT](https://www.youtube.com/watch?v=DeYTBDQnQZw)), that is, before being shipped to you. The OS’s job is to execute that precompiled binary code, and so you run your games.
+
+Now, `Super Mario 64` on the other hand runs `just-in-time` (JIT), maybe due to performance or portability reasons. 
+The emulator (Super Mario 3D All-Stars in this case, not yuzu) loads a native Nintendo 64 ROM of Super Mario 64, a JIT compiler takes the ROM and translates the original [MIPS](https://en.wikipedia.org/wiki/MIPS_architecture) (the architecture of the Nintendo 64’s CPU) instructions into [AArch64](https://en.wikipedia.org/wiki/AArch64) (the Switch’s CPU architecture) instructions in real time. 
+Just then the operating system will execute the game.
+
+{{< imgs
+	"./jitsrv.png| "
+  >}}
+
+This is similar to how yuzu translates AArch64 instructions into AMD64 ones, with the use of Dynarmic.
+
+This JIT is a functionality that yuzu didn’t have implemented, simply because no other game required it… or needed it.
```

Comment:
```suggestion
The JIT service, which is required to use JIT compilation on retail titles, is a functionality that yuzu didn’t have implemented, simply because no other game had ever needed it.
Additionally, there were some obstacles to implementing it in a direct way, since it requires calling custom code supplied by the game, something which was never needed by any previous service implementation.
```